org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -727,28 +727,17 @@ public static IJavaElement findElementAtSelection(ITypeRoot unit, int line, int
 				try {
 					offset = document.getLineOffset(line) + column;
 					if (offset > -1) {
+						if (offset > 0 && !isJavaIdentifierOrPeriod(contents.charAt(offset))) {
+							offset = offset - 1;
+						}
 						String name = parse(contents, offset);
 						if (name == null) {
 							return null;
 						}
-						SearchPattern pattern = SearchPattern.createPattern(name, IJavaSearchConstants.TYPE,
-								IJavaSearchConstants.DECLARATIONS, SearchPattern.R_FULL_MATCH);
-
-						IJavaSearchScope scope = createSearchScope(unit.getJavaProject(), preferenceManager);
-
-						List<IJavaElement> elements = new ArrayList<>();
-						SearchRequestor requestor = new SearchRequestor() {
-							@Override
-							public void acceptSearchMatch(SearchMatch match) {
-								if (match.getElement() instanceof IJavaElement) {
-									elements.add((IJavaElement) match.getElement());
-								}
-							}
-						};
-						SearchEngine searchEngine = new SearchEngine();
-						searchEngine.search(pattern,
-								new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() }, scope,
-								requestor, null);
+						List<IJavaElement> elements = search(unit, name, preferenceManager);
+						if (elements.isEmpty()) {
+							elements = search(unit.getJavaProject(), name, preferenceManager);

```

Comment:
The search engine is a long running job, it's better to pass the external progress monitor to make it cancellable.

A typical scenario is hover operation. When you move mouse on the editor, it will trigger multiple hover requests, and the previous unfinished request will be cancelled by client. So you'd better to pass the external monitor to the search engine.

Reply:
Fixed.